File: curse_of_vectorization/distances_among_pairs.jl
# Daniel C. Moura
# https://towardsdatascience.com/freeing-the-data-scientist-mind-from-the-curse-of-vectorization-11634c370107?source=friends_link&sk=e7b0086f3678261cb64b6dbb43744613
# linkedin.com/in/dmoura
# twitter.com/daniel_c_moura

# running this file:
# julia --optimize=3 --check-bounds=no distances_among_pairs.jl

using CPUTime
using LinearAlgebra

### functions under evaluation ###

function distances_among_pairs_loop(v)
    nin = length(v)
    nout = div(nin * nin - nin, 2) #length of the output vector
    dists = Vector{eltype(v)}(undef,nout) #preallocating output vector
    k = 1 #current position in the output vector
    @inbounds for i in 1:(nin-1)
        a = v[i]
        @inbounds for j in (i+1):nin
            b = v[j]
            dists[k] = abs(a-b)
            k += 1
        end
    end
    dists
end

distances_among_pairs_comp(v) =
    [abs(v[i]-v[j]) for i in eachindex(v), j in eachindex(v) if i<j]

function distances_among_pairs_outer(v)
    dists = v .- v'
    abs.(dists[tril!(trues(size(dists)), -1)])
end


### aux functions ###

function bench(label, f, v)
    GC.gc() #cleaning the garbage before each measurement
    print("$label:\n  ")
    @time @CPUtime f(v)
    println()
end

function readvec(run)
    a = Float64[]
    f = open("vec$run.csv", "r")
    for line in eachline(f)
        push!(a, parse(Float64, line))
    end
    close(f)
    a
end


### main ###

function main()
    println("Warming up...")
    v = fill(1.0, 3)
    distances_among_pairs_loop(v)
    distances_among_pairs_comp(v)
    distances_among_pairs_outer(v)

    for r in 1:3
        println("\n----- RUN $r -----")
        v = readvec(r)
        bench("LOOP",           distances_among_pairs_loop,  v)
        bench("COMPREHENSION",  distances_among_pairs_comp,  v)
        bench("OUTER",          distances_among_pairs_outer, v)
    end
end

main()

File: curse_of_vectorization/distances_among_pairs.R
# Copyright (c) 2019 Daniel Moura
# https://towardsdatascience.com/freeing-the-data-scientist-mind-from-the-curse-of-vectorization-11634c370107?source=friends_link&sk=e7b0086f3678261cb64b6dbb43744613
# linkedin.com/in/dmoura
# twitter.com/daniel_c_moura

# running this file:
# R --vanilla --slave < distances_among_pairs.R

library(profmem)
library(Rcpp)

### functions under evaluation ###

# baseline
distances_among_pairs_dist <- function(v)
  as.vector(dist(v, method="manhattan"))

distances_among_pairs_loop <- function(v) {
  nin = length(v)
  nout = (nin * nin - nin) %/% 2 #length of the output vector
  dists = vector(mode=typeof(v), length=nout) #preallocating output vector
  k = 1 #current position in the output vector
  for (i in seq_len(nin-1)) {
    a = v[i]
    for (b in v[seq(i+1, nin)]) {
      dists[k] = abs(a-b)
      k = k + 1
    }
  }
  dists
}

distances_among_pairs_rep <- function(v) {
  row <- rep(1:length(v), each=length(v))  #e.g 1 1 1 2 2 2 3 3 3
  col <- rep(1:length(v), times=length(v)) #e.g 1 2 3 1 2 3 1 2 3
  lower_tri <- which(row > col) #e.g. (2,1) (3,1) (3,2)
  abs(v[row[lower_tri]] - v[col[lower_tri]])
}

distances_among_pairs_outer <- function(v) {
  dists <- outer(v, v, '-')
  abs(dists[lower.tri(dists)])
}

### aux functions ###

bench.cpu <- function(label, f, v) {
  gc()
  cat(paste0(label, ":\n"))
  cpu.time <- sum(system.time(f(v))[1:2])
  cat(paste("  elapsed CPU time: ", cpu.time, "seconds\n"))
}

bench.mem <- function(label, f, v) {
  gc()
  cat(paste0(label, ":\n"))
  mem.used = total(profmem(f(v))) / 1048576
  cat(paste("  allocated:", mem.used, "MiB\n"))
}

readvec <- function(run)
  scan(paste0("vec", run, ".csv"), sep='\n', what = numeric(), quiet = TRUE)


### main ###

main <- function() {
  #compile C++ function
  sourceCpp("distances_among_pairs.cpp")  
  
  cat("Warming up...\n")
  v = vector("numeric", 3)
  distances_among_pairs_dist(v)
  distances_among_pairs_loop(v)
  distances_among_pairs_rep(v)
  distances_among_pairs_outer(v)
  distances_among_pairs_cpp(v)
  
  for (r in 1:3) {
    cat(paste("\n----- RUN", r, "-----\n"))
    v = readvec(r)
    bench.cpu("DIST",   distances_among_pairs_dist,  v)
    bench.cpu("LOOP",   distances_among_pairs_loop,  v)
    bench.cpu("REP",    distances_among_pairs_rep,   v)
    bench.cpu("OUTER",  distances_among_pairs_outer, v)
    bench.cpu("CPP",    distances_among_pairs_cpp,   v)
  }
  
  cat("\n----- MEMORY -----\n")
  bench.mem("DIST",   distances_among_pairs_dist,  v)
  bench.mem("LOOP",   distances_among_pairs_loop,  v)
  bench.mem("REP",    distances_among_pairs_rep,   v)
  bench.mem("OUTER",  distances_among_pairs_outer, v)
  bench.mem("CPP",    distances_among_pairs_cpp,   v)
}

main()
